File: tbfirst-ai-python/app/services/inpaint_service.py
import logging

from fastapi import HTTPException

from app.utils.image.uri import decode_image_to_inline
from app.schemas.image import InpaintRequest, InpaintResponse
from app.providers.gemini import GeminiProvider
from app.services.model_chain import image_chain
from app.services.prompt_engine import build_inpaint_prompt

logger = logging.getLogger(__name__)

_gemini = GeminiProvider()
_PRO_PREFERRED = "gemini-3-pro-image-preview"


async def inpaint(req: InpaintRequest) -> InpaintResponse:
    """局部重绘核心业务逻辑：图像解码 → mask 分支 → prompt 构建 → Gemini 调用。"""
    has_mask_image = bool(req.mask_image)
    has_reference = bool(req.reference_image)
    parts: list[dict] = []

    try:
        parts.append(decode_image_to_inline(req.masked_image))
        if has_mask_image:
            parts.append(decode_image_to_inline(req.mask_image))
        if has_reference:
            parts.append(decode_image_to_inline(req.reference_image))
    except ValueError as e:
        logger.error("[Inpaint] decode reference failed: %s", e)
        raise HTTPException(status_code=400, detail=str(e))

    prompt = build_inpaint_prompt(
        req.prompt, has_reference, req.negative_prompt, has_mask_image
    )
    parts.append({"text": prompt})

    preferred = req.model or _PRO_PREFERRED
    models = image_chain(preferred_first=preferred)

    try:
        result = await _gemini.generate_with_fallback(
            parts,
            config={
                "imageConfig": {"aspectRatio": req.aspect_ratio},
                "tools": [{"googleSearch": {}}],
            },
            models=models,
        )
    except Exception as e:
        logger.error("[Inpaint] failed: %s", e)
        raise HTTPException(status_code=502, detail=str(e))

    urls = result.get("urls", [])
    if not urls:
        raise HTTPException(status_code=502, detail="Inpainting generation failed.")

    return InpaintResponse(
        url=urls[0],
        used_model=result.get("used_model"),
        used_fallback=result.get("used_fallback", False),
    )

File: tbfirst-frontend/features/SmartInpaint/InpaintService.ts
/**
 * SmartInpaint — 对接 tbfirst 后端 /api/image/inpaint
 */
import { api } from '@/services/api';

interface InpaintResponse {
  url: string;
  jobId: number;
}

export const generateInpaintImage = async (
  maskedImageBase64: string,
  promptText: string,
  referenceImageBase64: string | null | undefined,
  aspectRatio: string = '3:4',
  /**
   * V5.XVII.B.9：二值 mask 图（黑=保留 / 白=重绘）。
   * 空 / undefined → 不进 body，后端走 V5.XVII Sprint A 单图路径（向后兼容）。
   * 非空 → 后端走双图路径（build_inpaint_prompt has_mask_image=True 模板）。
   */
  maskImageBase64?: string,
  /**
   * V5.XVII.6：绝对禁止段（多行字符串，每行一条）。
   * 空 / undefined → 不进 body，后端 prompt 输出与 V5.XVI 字节级一致。
   */
  negativePrompt?: string,
): Promise<string> => {
  const body: Record<string, unknown> = {
    maskedImage: maskedImageBase64,
    prompt: promptText,
    referenceImage: referenceImageBase64 || undefined,
    aspectRatio,
  };
  if (maskImageBase64) {
    body.maskImage = maskImageBase64;
  }
  if (negativePrompt && negativePrompt.trim()) {
    body.negativePrompt = negativePrompt;
  }
  const resp = await api.post<InpaintResponse>('/api/image/inpaint', body);
  return resp.url;
};
